src/demangle.cpp

```diff
@@ -117,14 +60,11 @@ void Demangle::cutArguments(char* s) {
 }
 
 char* Demangle::demangle(const char* s, bool full_signature) {
-    // Check if the mangled symbol ends with a Rust hash "17h<hex>E"
-    const char* e = strrchr(s, 'E');
-    if (e != NULL && e - s > 22 && e[-19] == '1' && e[-18] == '7' && e[-17] == 'h') {
-        const char* h = e - 16;
-        while ((*h >= '0' && *h <= '9') || (*h >= 'a' && *h <= 'f')) h++;
-        if (h == e) {
-            return demangleRust(s + 3, e - 19);
-        }
+    // try to demangle as Rust
+    struct demangle demangle;
+    rust_demangle_demangle(s, &demangle);
+    if (rust_demangle_is_known(&demangle)) {
+        return demangleRust(s, &demangle);
```

Comment:
Why is demangling performed in two steps? Can be the logic encapsulated in one call?

Reply:
because that's how it's done in the Rust code and the C code is a copy of the Rust code

Reply:
Do we want to keep this similarity? In any case, I'd expect to have this logic encapsulated in `demangleRust` rather than in a more generic `Demangle::demangle`.

Reply:
> Do we want to keep this similarity? In any case, I'd expect to have this logic encapsulated in demangleRust rather than in a more generic Demangle::demangle.

The code being this way means we don't try to demangle invalid Rust symbols as C++. I think that is correct.